### DISCUSSION TITLE
Pydantic v2 native solution for appending to list field of existing model instance if other model field value exists?

### DISCUSSION THREAD

**Discussion by sdruskat** [pydantic/pydantic#6464] [Question] [answered]
I'm trying to find a solution to do the following as a pydantic "native solution", e.g., through an `Annotated` validator or similar.

When adding a new `Thing` T2 to `MyModel.things`, I want to check if `MyModel.things` already contains a `Thing` with the same `id_field` as T2. If there is already a thing with the same id_field value, I'd like to add T2's `contents` to the `contents` of the existing `Thing`. If not, I want to append T2 to the list of `Thing`s in `MyModel`.

I'm building the model step by step, so `Thing`s are added to `MyModel.things` one by one.  
Currently, I'm using a function in the `MyModel` class (`add_thing(self, thing: Thing)`) to achieve this.

However, I'd like to ideally use functionality that `pydantic` provides instead. I'd like to be able to do this check and operation on any existing `Thing`s in either of these two ways. Or, if this isn't possible, in any other way:

```python
m = MyModel()

# Preferreably, add things like this and run check/operation on list
m.things.append(Thing(id_field="1", contents=["c1"]))
m.things.append(Thing(id_field="1", contents=["c2"]))
# Result: model has 1 Thing with a list of 2 contents

# Alternatively, building an interim list of things, then checking wouod also be fine:
things = []
things.append(Thing(id_field="1", contents=["c1"]))
things.append(Thing(id_field="1", contents=["c2"]))
m = MyModel(things=things)
# Result: model has 1 Thing with a list of 2 contents
```

The model looks as follows:

```python
@pydantic.dataclass
class Thing:
    id_field: str
    contents: list[str]


class MyModel(BaseModel):
    things: list[Thing] = Field([], title='The list of things')

    def add_thing(self, new_thing: Thing):
        for existing_thing in self.things:
            if existing_thing.id_field == new_thing.id_field:
                existing_thing.contents += new_thing.contents
                return
        self.things.append(new_thing)
```

The following test passes:

```python
def test_things():
    m = MyModel()
    thing = Thing(id_field="1", contents=["c1"])
    other_thing = Thing(id_field="1", contents=["c2"])
    m.add_thing(thing)
    m.add_thing(other_thing)
    assert len(m.things) == 1
    contents = m.things[0].contents
    assert len(contents) == 2
    ids = [i for i in contents]
    assert "c1" in ids
    assert "c2" in ids
```

Thanks for any pointers!

**Comment by ahmedabdou14** [comment]
Your work around is actually pretty good. If you want to validate `Thing`s when appending directly into the list without using an external method, you can checkout this [stackoverflow solution](https://stackoverflow.com/questions/67748856/how-to-validate-list-items-when-they-change-in-a-pydantic-model).

Side note:
```python
# things list will share the same reference for all instances of MyModel!
things: list[Thing] = Field([], title='The list of things')

# instead do this
things: list[Thing] = Field(default_factory=list, title='The list of things')
```

**Reply by sdruskat** [reply]
Thanks @Ahmed-Abdou14!

Can I assume that you mean [this solution](https://stackoverflow.com/a/67831328) (there are two)?

Thanks also for the sidenote, fixed it :).

**Reply by ahmedabdou14** [reply, marked as answer]
Yes i mean [this solution](https://stackoverflow.com/a/67831328/19517403) as you have mentioned.
This solution can be helpful if you want to use `things.append()` and add your own custom logic to it.
Another way is to create another custom class that extends `list` and you override the `__add__` method too,this would give you all features/methods of a normal list
